Geometry/Surfaces/QuadSurfaces/cone_class.f90

```diff
@@ -0,0 +1,451 @@
+
+    ! Along the axis
+    aabb(self % axis) = self % vertex(self % axis) + self % hMin
+    aabb(self % axis + 3) = self % vertex(self % axis) + self % hMax
+
+    ! On the plane of the basis
+    maxRadius = max(abs(self % hMin), abs(self % hMax)) * sqrt(self % tanSquare)
+    aabb(self % plane)     = self % vertex(self % plane) - maxRadius
+    aabb(self % plane + 3) = self % vertex(self % plane) + maxRadius
+
+  end function boundingBox
+
+  !!
+  !! Evaluate surface expression c = F(r)
+  !!
+  !! See surface_inter for details
+  !!
+  pure function evaluate(self, r) result(c)
+    class(cone), intent(in)                 :: self
+    real(defReal), dimension(3), intent(in) :: r
+    real(defReal)                           :: c, cMin, cMax, cBasis
+    real(defReal), dimension(3)             :: diff
+
+    ! Vector for the difference between position provided and vertex
+    diff = r - self % vertex
+
+    ! Evaluate the expression for the surface of the cone
+    c = dot_product(diff(self % plane), diff(self % plane)) - &
+        self % tanSquare * diff(self % axis) ** 2
+
+    ! Evaluate the expressions for the two basis of the cone and find the maximum
+    cMin = - diff(self % axis) + self % hMin
+    cMax = diff(self % axis) - self % hMax
+    cBasis = max(cMin, cMax)
+
+    ! Find the overall maximum
+    c = max(c, cBasis)
+
+  end function evaluate
+
+  !!
+  !! Return distance to the surface
+  !!
+  !! See surface_inter for details
+  !!
+  !! Solves quadratic intersection equation
+  !!   a*d^2 + 2b*d + c = 0
+  !!   c = cone surface expression
+  !!   b = (r1 - v1)u1 + (r2 - v2)u2 - t^2(r3 - v3)u3
+  !!   a = u1^2 + u2^2 - t^2 * u3^2
+  !!
+  pure function distance(self, r, u) result(dist)
+    class(cone), intent(in)                 :: self
+    real(defReal), dimension(3), intent(in) :: r
+    real(defReal), dimension(3), intent(in) :: u
+    real(defReal)                           :: dist
+    real(defReal), dimension(3)             :: diff, orientation
+    real(defReal)                           :: a, b, c, delta, d1, d2, near, far, &
+                                               test_near, test_far
+    logical(defBool)                        :: check
+    real(defReal), parameter :: FP_MISS_TOL = ONE + 10.0_defReal * epsilon(ONE)
+
+    ! Vector for the difference between position provided and vertex
+    diff = r - self % vertex
+
+    ! Explicitly specify cone orientation
+    orientation(self % plane) = ZERO
+    orientation(self % axis)  = sign(ONE, self % hMin)
+
+    ! Calculate quadratic components in the cone
+    a = dot_product(u(self % plane), u(self % plane)) - &
+        self % tanSquare * u(self % axis)**2
+    b = dot_product(diff(self % plane) , u(self % plane)) - &
+        self % tanSquare * diff(self % axis) * u(self % axis)
+    c = dot_product(diff(self % plane), diff(self % plane)) - &
+        self % tanSquare * diff(self % axis) ** 2
+
+    ! Calculate delta/4
```

Comment:
What is delta? You mean the discriminant? That also does not appear to be what is calculated here - that would be b^2 / 4 - a*c

Reply:
Sure, but it isn't clear what delta is. If you say 'calculate delta/4' there is no way for someone reading afterwards to understand what you are referring to.

Reply:
Should I call it discriminant instead? Is this what you mean?

Reply:
Yes, that would be easier to follow.